### PR TITLE
Implement getting storage slots deltas.

### PR DESCRIPTION
**Pull request by kayibal**
We'll use this when we get a revert.

### CODE REVIEW

File: migrations/2023-08-15-154244_tycho_initial_setup/up.sql [hunk 1]
```diff
@@ -387,15 +387,23 @@ create table if not exists contract_storage (
...
 	-- the contract this entry refers to.
 	"account_id" bigint references account(id) not null,
 	
 	-- the transaction that modified the slot to this entry.
 	"modify_tx" bigint references "transaction"(id) not null,
+
+	-- this is redundant with transaction.index, but included 
+	--	for performance reasons. Orders entries that are valid 
+	--	within the same version.
+	"ordinal" bigint not null,
```
Comment (carloszanella) [inline migrations/2023-08-15-154244_tycho_initial_setup/up.sql:406]: isn't this missing on ORM?

File: src/extractor/evm/mod.rs [hunk 2]
```diff
@@ -84,4 +84,4 @@ impl NormalisedMessage for AccountUpdate {
 }
 
 pub type EVMStateGateway<DB> =
-    StateGatewayType<DB, Block, Transaction, ERC20Token, SwapPool, Account, U256, U256>;
+    StateGatewayType<DB, Block, Transaction, ERC20Token, SwapPool, Account, H160, U256, U256>;
```
Comment (carloszanella) [inline src/extractor/evm/mod.rs:87]: why was this needed?

File: src/storage/mod.rs [hunk 3]
```diff
@@ -538,10 +548,11 @@ pub trait ContractStateGateway {
     ///     can't be located in storage.
     async fn get_slots_delta(
         &self,
-        id: ContractId,
+        id: Chain,
```
Comment (carloszanella) [inline src/storage/mod.rs:551]: why did you change this to be per chain instead of address?

File: src/storage/postgres/contract_state.rs [hunk 4]
```diff
@@ -70,14 +74,153 @@ where
 
     async fn get_slots_delta(
```
Comment (carloszanella) [inline src/storage/postgres/contract_state.rs:75]: should we already start adding docs to the public methods? or do you get documentation from the trait?

File: src/storage/postgres/contract_state.rs [hunk 5]
```diff
@@ -70,14 +74,153 @@ where
...
-            "id".to_owned(),
-        ))
+        target_version: Option<BlockOrTimestamp>,
+        conn: &mut AsyncPgConnection,
+    ) -> Result<HashMap<H160, HashMap<U256, U256>>, StorageError> {
+        let chain_id = self.get_chain_id(chain);
+        // To support blocks as versions, we need to ingest all blocks, else the
+        // below method can error for any blocks that are not present.
+        let start_version_ts = version_to_ts(&start_version, conn).await?;
+        let target_version_ts = version_to_ts(&target_version, conn).await?;
```
Comment (carloszanella) [inline src/storage/postgres/contract_state.rs:86]: This versioning is separate from the one of the extractors? Asking because there I think you defined server?

File: src/extractor/evm/mod.rs (hunk 2 above)
Comment (kayibal) [reply src/extractor/evm/mod.rs:87]: Because slots_deltas are returned for the whole chain now, the return type still needs to group them by account and I need something that serves as an account id here.

File: src/storage/mod.rs (hunk 3 above)
Comment (kayibal) [reply src/storage/mod.rs:551]: Because it is more sensible. There is no point of getting the delta changes of a single contract. Usually the chain reverts, not a single contract.

File: src/storage/postgres/contract_state.rs (hunk 4 above)
Comment (kayibal) [reply src/storage/postgres/contract_state.rs:75]: I think ideally things should be documented on the trait.

File: src/storage/postgres/contract_state.rs (hunk 5 above)
Comment (kayibal) [reply src/storage/postgres/contract_state.rs:86]: I think I wrote a big documentation about versioning here. The extractors `version` attribute does not achieve table versioning btw, it just records which version of the extractor substreams package we want to run.

File: migrations/2023-08-15-154244_tycho_initial_setup/up.sql (hunk 1 above)
Comment (kayibal) [reply migrations/2023-08-15-154244_tycho_initial_setup/up.sql:406]: I'll add this in my next PR.
